File: placement-flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/common/data/widgets/CanvasData.kt
package com.appsamurai.storyly.storyly_placement_flutter.common.data.widgets

import com.appsamurai.storyly.canvas.data.CanvasDataPayload
import com.appsamurai.storyly.canvas.data.model.CanvasItem
import com.appsamurai.storyly.canvas.data.model.CanvasPayload
import com.appsamurai.storyly.core.STRWidgetType
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.encodeSTRProductItem

fun encodeCanvasDataPayload(data: CanvasDataPayload): Map<String, Any?> {
    return mapOf(
        "type" to STRWidgetType.Canvas.raw,
        "items" to data.items.map { encodeCanvasItem(it) }
    )
}

fun encodeCanvasPayload(payload: CanvasPayload?): Map<String, Any?>? {
    payload ?: return null
    return mapOf(
        "item" to encodeCanvasItem(payload.item)
    )
}

fun encodeCanvasItem(item: CanvasItem?): Map<String, Any?>? {
    item ?: return null
    return mapOf(
        "actionUrl" to item.actionUrl,
        "actionProducts" to item.actionProducts?.map { encodeSTRProductItem(it) }
    )
}

File: placement-flutter/storyly_placement_flutter/ios/Classes/Common/Data/Widgets/CanvasData.swift
import StorylyCore
import StorylyCanvas


func encodeCanvasDataPayload(_ data: CanvasDataPayload) -> [String: Any] {
    return [
        "type": STRWidgetType.canvas.description,
        "items": data.items.map { encodeCanvasItem($0) }
    ]
}

func encodeCanvasPayload(_ payload: CanvasPayload?) -> [String: Any]? {
    guard let payload = payload else { return nil }
    
    return [
        "item": encodeCanvasItem(payload.item)
    ].compactMapValues { $0 }
}

func encodeCanvasItem(_ item: CanvasItem?) -> [String: Any?]? {
  guard let item = item else { return nil }
  return [
    "actionUrl": item.actionUrl,
    "actionProducts": item.actionProducts?.map { encodeSTRProductItem($0) }
  ]
}
